RecoLocalTracker/SiStripRecHitConverter/interface/StripCPE.h

```diff
@@ -28,6 +29,45 @@ class StripCPE : public StripClusterParameterEstimator
 	    const SiStripLatency&);    
   LocalVector driftDirection(const StripGeomDetUnit* det) const;
 
+ struct Param {
+    Param() : topology(nullptr) {}
+    StripTopology const * topology;
+    LocalVector drift;
+    float thickness, invThickness,pitch_rel_err2, maxLength;
+    int nstrips;
+    float backplanecorrection;
+    SiStripDetId::ModuleGeometry moduleGeom;
+    float coveredStrips(const LocalVector&, const LocalPoint&) const;
+  };
+
+  
+  struct AlgoParam {
+    Param const & p; const LocalTrajectoryParameters & ltp;
+    SiStripDetId::SubDetector loc; float afullProjection; float corr;
+  };
+
+
+  virtual StripClusterParameterEstimator::LocalValues
+  localParameters( const SiStripCluster& cl, AlgoParam const & ap) const {
+    return std::make_pair(LocalPoint(), LocalError());
+  }
+  
+  AlgoParam getAlgoParam(const GeomDetUnit& det, const LocalTrajectoryParameters & ltp) const {
+
+    StripCPE::Param const & p = param(det);
+    SiStripDetId::SubDetector loc = SiStripDetId( det.geographicalId() ).subDetector();  
+ 
+    LocalVector track = ltp.momentum();
+    track *= -p.thickness/track.z();
```

Comment:
On 5/6/16 8:44 AM, Vincenzo Innocente wrote:
> In RecoLocalTracker/SiStripRecHitConverter/interface/StripCPE.h
> <https://github.com/cms-sw/cmssw/pull/14344#discussion_r62294072>:
> 
>> +    SiStripDetId::SubDetector loc; float afullProjection; float corr;
>> +  };
>> +
>> +
>> +  virtual StripClusterParameterEstimator::LocalValues
>> +  localParameters( const SiStripCluster& cl, AlgoParam const & ap) const {
>> +    return std::make_pair(LocalPoint(), LocalError());
>> +  }
>> +  
>> +  AlgoParam getAlgoParam(const GeomDetUnit& det, const LocalTrajectoryParameters & ltp) const {
>> +
>> +    StripCPE::Param const & p = param(det);
>> +    SiStripDetId::SubDetector loc = SiStripDetId( det.geographicalId() ).subDetector();  
>> + 
>> +    LocalVector track = ltp.momentum();
>> +    track *= -p.thickness/track.z();
> 
> I think we need to have a more in depth discussion about "FPE" and NaN
> particularly in the context of vectorization and the future move to
> vector hardware. We cannot afford to protect each single operation. The
> current accepted wisdom is to let NaN to propagate and trap it at very
> high level. 

"A very high level" better be the output of the module or at worst
output of a sequence of modules
that has consumable output downstream.

The problem is more complex for utility/tools which are used in many places.
A fraction of users of the code may not care about execution speed much.
Maybe for that case it's practical to add the checks to avoid FPE
(template to hand over FPE-safe and unsafe interface?).

> Maybe we need to invite an expert and give us a lecture on
> how one manage this type of issues in HPC.

Do you have a name in mind?
Maybe send a suggestion by email.

> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/14344/files/e77be1aadab9d37f1baf957e190b179f95a1fc39#r62294072>
>